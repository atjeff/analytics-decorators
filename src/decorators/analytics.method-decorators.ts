import { sendEvent } from '../analytics.core';
import { AnalyticsEventFactory } from '../types/analytics-event.model';

export function TrackAnalyticsEvent<T = any>(event: AnalyticsEventFactory<T>): MethodDecorator {
    return (target: object, key: string, descriptor: PropertyDescriptor) => {
        const original = descriptor.value;

        descriptor.value = function(...args: any[]) {
            const analyticsEvent = typeof event === 'function' ? event(this) : event;

            if (analyticsEvent) {
                sendEvent(analyticsEvent);
            }

            return original.apply(this, args);
        };

        return descriptor;
    };
}
