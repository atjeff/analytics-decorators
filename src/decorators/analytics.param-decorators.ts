import { get, getMany } from '../analytics.core';

/**
 * Sets specified field values into an object from GA tracker.
 *
 * @export
 * @param {string[]} fieldNames name from GA tracker
 * @returns {ParameterDecorator}
 */

// This isn't going to work, you can't change the values with a param decorator, only flag them with metadata to be replaced by a method decorator later.
export function Field(fieldNames?: string | string[]): ParameterDecorator {
    /**
     * @param {Object} target The prototype of the class being decorated.
     * @param {string} key The name of the method that contains the parameter being decorated.
     * @param {*} descriptor The index of that parameter being decorated.
     */
    return (target: object, key: string, descriptor: any) => {
        if (Array.isArray(fieldNames)) {
            descriptor.value = getMany(fieldNames);
        } else if (fieldNames) {
            descriptor.value = get(fieldNames);
        }

        return descriptor;
    };
}
