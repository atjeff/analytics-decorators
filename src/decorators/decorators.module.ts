export * from './analytics.method-decorators';
export * from './tracker-field.decorators';
