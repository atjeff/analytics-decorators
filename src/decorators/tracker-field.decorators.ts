import 'reflect-metadata';
import { get, getMany } from '../analytics.core';
import { FieldDecoratorsMetadata } from './models/field-decorators-metadata.model';

/**
 * Creates metadata key using propertyName
 *
 * @param {string} propertyName
 */
const createTrackerMetadataKey = (propertyName: string) => `__tracker_${propertyName}_parameters`;

/**
 * Decorates a method, replaces @Field() decorated params with the passed in tracker field value
 *
 * @export
 * @returns {MethodDecorator}
 */
export function UseTrackerFields(): MethodDecorator {
    return (target: any, propertyName: string, descriptor: PropertyDescriptor) => {
        const originalMethod = descriptor.value;

        descriptor.value = function(...args: any[]) {
            const metadataKey = createTrackerMetadataKey(propertyName);

            const requiredParameters: FieldDecoratorsMetadata[] = Reflect.getOwnMetadata(metadataKey, target, propertyName) || [];

            const modifiedArgs = requiredParameters.reduce((accumulator, { parameterIndex, fieldNames }) => {
                const trackerValue = Array.isArray(fieldNames) ? getMany(fieldNames) : get(fieldNames);

                if (trackerValue) {
                    accumulator[parameterIndex] = trackerValue;
                }

                return accumulator;
            }, args.slice());

            return originalMethod.apply(this, modifiedArgs);
        };

        return descriptor;
    };
}

/**
 * Sets tracker field value on function param
 *
 * @export
 * @param {(string | string[])} [fieldNames]
 * @returns {ParameterDecorator}
 */
export function Field(fieldNames?: string | string[]): ParameterDecorator {
    return (target: Object, propertyName: string, parameterIndex: number) => {
        const metadataKey = createTrackerMetadataKey(propertyName);
        const existingRequiredParameters: FieldDecoratorsMetadata[] = Reflect.getOwnMetadata(metadataKey, target, propertyName) || [];

        existingRequiredParameters.push({ parameterIndex, fieldNames });

        Reflect.defineMetadata(metadataKey, existingRequiredParameters, target, propertyName);
    };
}
