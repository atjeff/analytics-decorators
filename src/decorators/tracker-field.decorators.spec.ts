jest.mock('../analytics.core.ts');

import * as analyticsCore from '../analytics.core';
import { Field, UseTrackerFields } from './tracker-field.decorators';

describe('Tracker Field Decorators', () => {
    describe('UseTrackerFields', () => {
        it(`should not modify any arguments if they don't have metadata`, () => {
            const args = ['test1', 'test2'];
            const calledSpy = jest.fn();

            class TestUseTrackerFields {
                @UseTrackerFields()
                testMethod(arg1: string, arg2: string) {
                    calledSpy(arg1, arg2);
                }
            }

            new TestUseTrackerFields().testMethod(args[0], args[1]);

            expect(calledSpy).toMatchSnapshot();
        });

        it(`should replace a specified key param with its value`, () => {
            const args = ['test1', 'test2'];
            const fieldName = 'appId';
            const fieldValue = 'app-id-1234';
            const calledSpy = jest.fn();
            const getSpy = jest.spyOn(analyticsCore, 'get').mockReturnValue(fieldValue);

            class TestUseTrackerFields {
                @UseTrackerFields()
                testMethod(arg1: string, @Field(fieldName) arg2: string) {
                    calledSpy(arg1, arg2);
                }
            }

            new TestUseTrackerFields().testMethod(args[0], args[1]);

            expect(calledSpy).toMatchSnapshot();
            expect(getSpy).toHaveBeenCalled();

            getSpy.mockRestore();
        });

        it(`should replace an array of keys param with their values`, () => {
            const args = ['test1', 'test2'];
            const fieldNames = ['appId', 'appVersion'];
            const fieldValues = ['app-id-1234', '4.2.0'];
            const fieldsResponse = { [fieldNames[0]]: fieldValues[0], [fieldNames[1]]: fieldValues[1] };
            const calledSpy = jest.fn();
            const getManySpy = jest.spyOn(analyticsCore, 'getMany').mockReturnValue(fieldsResponse);

            class TestUseTrackerFields {
                @UseTrackerFields()
                testMethod(arg1: string, @Field(fieldNames) arg2) {
                    calledSpy(arg1, arg2);
                }
            }

            new TestUseTrackerFields().testMethod(args[0], args[1]);

            expect(calledSpy).toMatchSnapshot();
            expect(getManySpy).toHaveBeenCalled();
        });

        it(`shouldn't replace anything if get doesn't return values`, () => {
            const args = ['test1', 'test2'];
            const fieldName = 'appId';
            const calledSpy = jest.fn();

            class TestUseTrackerFields {
                @UseTrackerFields()
                testMethod(arg1: string, @Field(fieldName) arg2: string) {
                    calledSpy(arg1, arg2);
                }
            }

            new TestUseTrackerFields().testMethod(args[0], args[1]);

            expect(calledSpy).toMatchSnapshot();
        });
    });
});
