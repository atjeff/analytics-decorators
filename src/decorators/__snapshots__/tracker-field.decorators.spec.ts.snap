// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tracker Field Decorators UseTrackerFields should not modify any arguments if they don't have metadata 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "test1",
      "test2",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Tracker Field Decorators UseTrackerFields should replace a specified key param with its value 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "test1",
      "app-id-1234",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Tracker Field Decorators UseTrackerFields should replace an array of keys param with their values 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "test1",
      Object {
        "appId": "app-id-1234",
        "appVersion": "4.2.0",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Tracker Field Decorators UseTrackerFields shouldn't replace anything if get doesn't return values 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "test1",
      "test2",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
